migrations/versions/fc34137ad3ba_.py

```diff
@@ -0,0 +1,29 @@
+"""Rename paid column to free
+
+Revision ID: fc34137ad3ba
+Revises: 205742d3b3f5
+Create Date: 2020-09-30 22:35:20.007768
+
+"""
+from alembic import op
+import sqlalchemy as sa
+import sqlalchemy_utils
+
+
+# revision identifiers, used by Alembic.
+revision = 'fc34137ad3ba'
+down_revision = '205742d3b3f5'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.alter_column('resource', 'paid', new_column_name='free', existing_type=sa.Boolean, nullable=False)
```

Comment:
After running the upgrade, I got this error:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.NotNullViolation: column "free" contains null values


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/flask", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/flask/cli.py", line 967, in main
    cli.main(args=sys.argv[1:], prog_name="python -m flask" if as_module else None)
  File "/usr/local/lib/python3.7/site-packages/flask/cli.py", line 586, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/flask/cli.py", line 426, in decorator
    return __ctx.invoke(f, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/cli.py", line 134, in upgrade
    _upgrade(directory, revision, sql, tag, x_arg)
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/__init__.py", line 96, in wrapped
    f(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/flask_migrate/__init__.py", line 271, in upgrade
    command.upgrade(config, revision, sql=sql, tag=tag)
  File "/usr/local/lib/python3.7/site-packages/alembic/command.py", line 298, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.7/site-packages/alembic/script/base.py", line 489, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/usr/local/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.7/site-packages/alembic/util/compat.py", line 184, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "migrations/env.py", line 89, in <module>
    run_migrations_online()
  File "migrations/env.py", line 82, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/usr/local/lib/python3.7/site-packages/alembic/runtime/environment.py", line 846, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/local/lib/python3.7/site-packages/alembic/runtime/migration.py", line 522, in run_migrations
    step.migration_fn(**kw)
  File "/src/migrations/versions/ed6f77fb8240_.py", line 22, in upgrade
    op.add_column('resource', sa.Column('free', sa.Boolean(), nullable=False))
  File "<string>", line 8, in add_column
  File "<string>", line 3, in add_column
  File "/usr/local/lib/python3.7/site-packages/alembic/operations/ops.py", line 1929, in add_column
    return operations.invoke(op)
  File "/usr/local/lib/python3.7/site-packages/alembic/operations/base.py", line 373, in invoke
    return fn(self, operation)
  File "/usr/local/lib/python3.7/site-packages/alembic/operations/toimpl.py", line 132, in add_column
    operations.impl.add_column(table_name, column, schema=schema, **kw)
  File "/usr/local/lib/python3.7/site-packages/alembic/ddl/impl.py", line 238, in add_column
    self._exec(base.AddColumn(table_name, column, schema=schema))
  File "/usr/local/lib/python3.7/site-packages/alembic/ddl/impl.py", line 141, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1073, in _execute_ddl
    compiled,
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1317, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1511, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) column "free" contains null values

[SQL: ALTER TABLE resource ADD COLUMN free BOOLEAN NOT NULL]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```

I'm not sure why :-(

Reply:
https://github.com/OperationCode/resources_api/blob/b246f3ae766aa924450ffe32e7a5956471e64db6/migrations/versions/514a36eb1161_.py#L37

Reply:
I had thought that even though we had nullable set to true by a migration error we wouldnt have any data in the db that had paid as null - I think you could go into a flask shell and query any paid=None resource and update them to true instead then run upgrade

Reply:
Or wait.. In the traceback:
```
File "/src/migrations/versions/ed6f77fb8240_.py", line 22, in upgrade
    op.add_column('resource', sa.Column('free', sa.Boolean(), nullable=False))
```
This shouldnt be right, it seems to have auto generated a migration script (different than the one I added) which is incorrect

Reply:
Try deleting that auto generated script and run the current migration files and see if it applied our migration file - if it doesnt then I guess we should run the generation of migration command and edit the new migration file to be the same commands as the one we had then run migrate

Reply:
I can't even find any resources where this is a thing
```python
>>> r = Resource.query.filter(and_(Resource.paid != False, Resource.paid != True)).all()
>>> len(r)
0
```

Reply:
yeah I think I encountered this when I was developing locally, the default migration that alembic generates when removing the paid field from sqlalchemy model and adding free field is adding a new column to the db instead of renaming which isnt what we want and its the reason of the error thats popping up - its why I had to alter the migration script to rename. if you see the traceback you will notice that `ed6f77fb8240_.py` isnt one of our migration files that could be found [here](https://github.com/OperationCode/resources_api/tree/main/migrations/versions). I am not sure why alembic didn't take notice of our migration current migration files instead of generating its own

Reply:
Try running the following before running upgrade:
```
flask db stamp head
```
This should solve it I think

Reply:
Ok, so I did that and I realized that I think the app thought that the DB is on fc34137ad3ba already when the new container is deployed, but it's not. So the solution was to do `flask db stamp 205742d3b3f5` and then `flask db upgrade` which seemed to fix the problem.

The only problem with this now is that in trying to fix it earlier, I changed `free` to be nullable, and I don't really want that, especially since all of the current resources are non-null.

So I'm going to change that, and then repeat this process for prod and hopefully everything will work out 🤞

Reply:
An apparently successful deploy to prod just now. This time when I logged in this is what happened:
```
uwsgi@resources-api-58d75f6b89-f8v9n:/src$ flask db current
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
205742d3b3f5
uwsgi@resources-api-58d75f6b89-f8v9n:/src$ flask db upgrade
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 205742d3b3f5 -> fc34137ad3ba, Rename paid column to free
```

So I guess I must have been doing something else to mess everything up before... Oh well, whatever, it's all resolved now. Thanks so much for your help @Kandeel4411 !!!

Reply:
Oh awesome 👍